### PR TITLE
Automated Boosterpasses

### LINKED ISSUE #15: Automate booster passes with the bot
### Feature request

**Checklist**
- [x] I have read and agree to the README (found on https://github.com/Minehut/Meta)
- [x] This is not a bug report, support request or plugin request
- [x] I have searched the Issue Tracker to make sure this is not a duplicate issue
- [x] This issue does not contain any sensitive information 

**Platform**

Discord

**Describe the feature**

Add a **!boosterpass** Discord bot command with the usage:
- `!boosterpass info [user]` - shows user's given and received passes
- `!boosterpass give <user>` - give one of your booster passes to that user, provided they don't already have one assigned to them
- `!boosterpass revoke <user>` - take your booster pass away from that user

Maximum passes you can assign would stay the same

**How the feature is useful**

The Minehut Discord bot should automate assigning booster passes so senior staff don't need to spend time managing it and users can reassign their passes instantly.

---
**Use 👍 and 👎 reactions to indicate you agree or disagree with this suggestion! Feel free to discuss the idea in the comments.**

### PR DESCRIPTION
**Pull request by Sniped**
Allow booster passes to be automated completely with the ability to grant and revoke booster passes if you're a nitro booster.

Closes https://github.com/Minehut/Meta/issues/15 

**Commands**
* `!boosterpass give <member>`
* `!boosterpass info <member>`
* `!boosterpass revoke <member>`

**Configuration Changes**
* Adds configuration for booster passes
  * Once the booster pass property is declared in the features configuration for a guild, it is automatically enabled.
  * `maximumGrantedBoosterPasses` - allows the number of maximum granted booster passes to be changed, by default it's 2.

**Other Changes**
* Adds `enforcePermissionLevel` permission to command options to ensure that the command executor is a nitro booster. 
* Commit b8ed83ae8a4419d4889a689689e828679458528f prettified some code outside of this PR's scope, however shouldn't have any noticeable changes.
* With the addition of the `enforcePermissionLevel` property, some functions pertaining to permission checks were put into their own files for the sake of organization.

### CODE REVIEW

**Review by theminecoder — CHANGES_REQUESTED** [review]
Instead of having the import as a bot command, could you please extract it into a file I can run on its own.

**Comment by BennyDoesStuff** [comment]
What if a user does not want a booster pass, is there a way for them to opt-out?

**Comment by Sniped** [comment]
> What if a user does not want a booster pass, is there a way for them to opt-out?

Never thought about that, depending on how many people want to opt out, maybe we can just keep a value in the configuration and update it when someone wants to opt out? We could also make it a command, however I imagine that would need another table to keep track of the people that are opted out.

**Comment by BennyDoesStuff** [comment]
hmm, I mention that because I do know of a few situations where I had to manually remove their booster pass because they were not interested in having it. We should either have a way to opt-out or a way to remove that from someone if they request

**Comment by Sniped** [comment]
> hmm, I mention that because I do know of a few situations where I had to manually remove their booster pass because they were not interested in having it. We should either have a way to opt-out or a way to remove that from someone if they request

Then I think we can just have a config value for people that don't want a booster pass. Do you know of any cases right now where a user has a booster pass even though they don't want it?

**Comment by theminecoder** [comment]
> > What if a user does not want a booster pass, is there a way for them to opt-out?
> 
> Never thought about that, depending on how many people want to opt out, maybe we can just keep a value in the configuration and update it when someone wants to opt out? We could also make it a command, however I imagine that would need another table to keep track of the people that are opted out.

For right now I think the current situation is fine as long as we have a command to remove the pass from people, but I wouldn't be against maybe moving to a central player document where we can store preferences/punishments/other things like this going forward.

**Comment by Sniped** [comment]
> Instead of having the import as a bot command, could you please extract it into a file I can run on its own.

The recent push I did should work with the root env file as long as the script is run in the same folder as it. I made it in reference to jellz's database migration script.

**Review by theminecoder — CHANGES_REQUESTED** [review]
Remove the current export file and we are good to go. I'll grab it when the time is right :)

File: src/command/boosterPass/boosterPass.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	constructor() {
+		super('boosterPass', {
+			aliases: ['boosterpass', 'bp'],
+			description: {
+				content: `Give, revoke, or retrieve info on booster passes
+                Available subcommands:
+                • **give** \`<member>\`
+                • **info** \`<member>\`
+                • **revoke** \`<member>\`
+                • **import**
```
Comment (theminecoder) [inline src/command/boosterPass/boosterPass.ts]: Left over?

File: src/command/boosterPass/boosterPass.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                `,
+				usage: '<method> <...arguments>',
+				examples: [
+					'give @Facto',
+					'give 535986058991501323',
+					'info @Facto',
+					'info 535986058991501323',
+					'revoke @Facto',
+					'revoke 535986058991501323',
+					'import',
```
Comment (theminecoder) [inline src/command/boosterPass/boosterPass.ts]: Left Over?

File: src/command/boosterPass/boosterPass.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+		});
+	}
+
+	*args() {
+		const method = yield {
+			type: [
+				['boosterpass-give', 'give'],
+				['boosterpass-revoke', 'revoke'],
+				['boosterpass-info', 'info'],
+				['boosterpass-import', 'import'],
```
Comment (theminecoder) [inline src/command/boosterPass/boosterPass.ts]: Left Over?

**Review by theminecoder — CHANGES_REQUESTED** [review]

File: package.json [hunk 4]
```diff
@@ -32,7 +31,8 @@
 		"numeral": "^2.0.6",
 		"parse-duration": "jellz/parse-duration",
 		"ts-node": "^8.8.2",
-		"typescript": "^3.8.3"
+		"typescript": "^3.8.3",
+		"xlsx": "^0.16.7"
```
Comment (theminecoder) [inline package.json]: Does this need to be here anymore?

```suggestion
```

Comment (Sniped) [reply package.json]: Not sure why I didn't remove that, must've thought I did last night.

File: package.json [hunk 5]
```diff
@@ -32,7 +31,7 @@
 		"numeral": "^2.0.6",
 		"parse-duration": "jellz/parse-duration",
 		"ts-node": "^8.8.2",
-		"typescript": "^3.8.3"
+		"typescript": "^3.8.3",
```
Comment (theminecoder) [inline package.json]:
```suggestion
		"typescript": "^3.8.3"
```

**Review by theminecoder — APPROVED** [review]

**Pull request merged**